File: 2021/javascript/day2/puzzle.js
let utils = require('../aoc_utils')
let data = require('./input')

function run(part) {
	const d = parseData(data);
	return (part === 1) ? utils.runWithTimer(partOne, d) : utils.runWithTimer(partTwo, d);
}

function parseData(data) {
	let d = [];
	data.trim().split('\n').forEach(line => {
		d.push(line.split(' '));
	});
	return d;
}

function partOne(data) {
	let dist = 0;
	let depth = 0;
	for (const d of data) {
		switch (d[0]) {
			case 'forward':
				dist += parseInt(d[1]);
				break;
			case 'down':
				depth += parseInt(d[1]);
				break;
			case 'up':
				depth -= parseInt(d[1]);
				break;
		}
	}
	return dist * depth;
}

function partTwo(data) {
	let dist = 0;
	let depth = 0;
	let aim = 0;
	for (const d of data) {
		switch (d[0]) {
			case 'forward':
				dist += parseInt(d[1]);
				depth += aim * parseInt(d[1]);
				break;
			case 'down':
				aim += parseInt(d[1]);
				break;
			case 'up':
				aim -= parseInt(d[1]);
				break;
		}
	}
	return dist * depth;
}

module.exports = { run, parseData, partOne, partTwo }

File: 2021/java/src/main/java/aoc/day2/Puzzle.java
package aoc.day2;

import aoc.AocPuzzle;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static aoc.utils.LineReader.readLines;

public class Puzzle extends AocPuzzle<List<Instruction>, Integer> {

	public static void main(String[] args) {
		new Puzzle().runWithTimers();
	}

	public Puzzle() {
		super();
	}

	public Puzzle(List<Instruction> instructions) {
		super(instructions);
	}

	@Override
	public Integer partOne() {
		Position p = new Position();
		for (Instruction d : input) {
			switch (d.direction()) {
				case "forward" -> p.moveForward(d.amount());
				case "down" -> p.moveDown(d.amount());
				case "up" -> p.moveUp(d.amount());
			}
		}
		return p.calculate();
	}

	@Override
	public Integer partTwo() {
		Position p = new Position();
		for (Instruction d : input) {
			switch (d.direction()) {
				case "forward" -> {
					p.moveForward(d.amount());
					p.moveDown(p.getAim() * d.amount());
				}
				case "down" -> p.aimDown(d.amount());
				case "up" -> p.aimUp(d.amount());
			}
		}
		return p.calculate();
	}


	@Override
	public List<Instruction> getInput() {
		List<String> lines = readLines(getInputLocation());
		List<Instruction> retVal = new ArrayList<>();
		for (String line : lines) {
			String[] parts = line.split(" ");
			retVal.add(new Instruction(parts[0], Integer.parseInt(parts[1])));
		}
		return retVal;
	}

	@Override
	public String getInputLocation() {
		return "day2/input.txt";
	}
}

class Position {
	private int horizontal;
	private int depth;
	private int aim;

	public void moveForward(int amount) {
		horizontal += amount;
	}

	public void moveDown(int amount) {
		depth += amount;
	}

	public void moveUp(int amount) {
		depth -= amount;
	}

	public void aimDown(int amount) {
		aim += amount;
	}

	public void aimUp(int amount) {
		aim -= amount;
	}

	public int calculate() {
		return horizontal * depth;
	}

	public int getAim() {
		return aim;
	}
}

record Instruction(String direction, int amount) {
}
